overlord/devicestate/devicestate.go

```diff
@@ -1205,11 +1222,34 @@ func Remodel(st *state.State, new *asserts.Model, localSnaps []*snap.SideInfo, p
 		return nil, err
 	}
 
+	prevRev, err := findRevisionForModel(st, new)
+	if err != nil {
+		return nil, err
+	}
+	if new.Revision() < prevRev {
+		return nil, fmt.Errorf("cannot remodel to older revision %d of model %s/%s than last revision %d known to the device", new.Revision(), new.BrandID(), new.Model(), prevRev)
+	}
+
+	// TODO: we need dedicated assertion language to permit for
+	// model transitions before we allow cross vault
+	// transitions.
+
+	remodelKind := ClassifyRemodel(current, new)
+
 	if _, err := findSerial(st, nil); err != nil {
+		cont := false
```

Comment:
I think you could simplify this a bit by inverting some conditions, although that might not convey the message as clearly:
```go
if _, err := findSerial(st, nil); err != nil {
	if !errors.Is(err, state.ErrNoState) {
		return nil, err
	}

	// it is allowed to remodel without serial for offline remodels that are
	// update only
	if len(localSnaps) == 0 || remodelKind != UpdateRemodel {
		return nil, fmt.Errorf("cannot remodel without a serial")
	}
}
```

Reply:
I went with something slightly different, the main issue with this switch to || !